ros/src/tl_detector/tl_detector.py

```diff
@@ -49,13 +51,20 @@ def __init__(self):
         self.last_wp = -1
         self.state_count = 0
 
+        self.waypoints_2d = None
+        self.waypoints_tree = None
+        self.n_waypoints = 0
         rospy.spin()
 
     def pose_cb(self, msg):
         self.pose = msg
 
     def waypoints_cb(self, waypoints):
         self.waypoints = waypoints
+        # Preprocess to make it faster to search for closest waypoint using the k-d tree algorithm.
```

Comment:
There is a class WaypointSearch in waypoint_updater.py that do exactly this type of search.

Reply:
Thanks, that's noted, I'll refactor that in a later step.